opendevin/controller/state/state.py

```diff
@@ -29,6 +29,8 @@ class State:
     root_task: RootTask = field(default_factory=RootTask)
     iteration: int = 0
     max_iterations: int = 100
+    max_iterations_per_task: int = 100
+    global_iteration: int = 0
```

Comment:
QQ: why `global_iteration=0` while `max_iterations_per_task=100`

Reply:
`global_iteration` is the global counter, not a global limit?

Reply:
'global_iteration` in the state is to track the global iteration count. Thank you for pointing it out, I have figured out a issue in my logic in code after reviewing it, and I was working on it.

Im changing the `config.py` to introduce a new configuration parameter `global_max_iterations`, which sets a global limit on the total number of iterations that can be executed by all agents and their delegates. This parameter is integrated into the application configuration to ensure it is available for controlling the overall iteration count across different agents.

Due to that in state.py, a new attribute `global_iteration` is added to the State class to track the total number of iterations executed globally. This attribute is incremented alongside the task-specific iteration in the `AgentController` to maintain a cumulative count of iterations. This tracking is essential for ensuring that the global iteration limit set in the configuration is respected across all agents and their delegates.

The `agent_controller.py` file incorporates the `global_max_iterations` and `global_iteration` to enforce the iteration limits. When a delegate agent is created, its max_iterations is set based on the remaining global iterations to ensure the total does not exceed the global limit. The `_step` method is updated to check both the task-specific and global iteration counts, stopping the agent if either exceeds the defined limits. These changes work together to cap the total iteration count by tracking and enforcing the limits at both the task and global levels.

Can I commit these changes? And also, Can I git checkout the `evaluation` folder files in this pr?

Reply:
Sure, please go ahead.